MotionPhotoReader/app/src/main/java/com/google/android/libraries/motionphotoreader/MotionPhotoReader.java

```diff
@@ -0,0 +1,302 @@
+        if (Build.VERSION.SDK_INT < M) {
+            Log.e("MotionPhotoReader", "Insufficient Android build version");
+            return;
+        }
+
+        lowResDecoder.setCallback(new MediaCodec.Callback() {
+
+            @Override
+            public void onInputBufferAvailable(@NonNull MediaCodec codec, int index) {
+                boolean result = availableInputBuffers.offer(index);
+                Log.d(TAG, "Input buffers: " + availableInputBuffers.toString());
+            }
+
+            @Override
+            public void onOutputBufferAvailable(@NonNull MediaCodec codec,
+                                                int index,
+                                                @NonNull MediaCodec.BufferInfo info) {
+                Bundle bufferData = new Bundle();
+                bufferData.putInt("BUFFER_INDEX", index);
+                bufferData.putLong("TIMESTAMP_US", info.presentationTimeUs);
+                boolean result = availableOutputBuffers.offer(bufferData);
+                Log.d(TAG, "Output buffers: " + availableOutputBuffers.toString());
+            }
+
+            @Override
+            public void onError(@NonNull MediaCodec codec, @NonNull MediaCodec.CodecException e) {
+                Log.e(TAG, "Error while decoding", e);
+            }
+
+            @Override
+            public void onOutputFormatChanged(@NonNull MediaCodec codec,
+                                              @NonNull MediaFormat format) {
+
+            }
+        }, mediaHandler);
+
+        lowResDecoder.configure(videoFormat, surface, null, 0);
+        lowResDecoder.start();
+    }
+
+    /**
+     * Sets up and starts a new handler thread for managing frame advancing calls and available buffers.
+     */
+    private void startBufferThread() {
+        bufferWorker = new HandlerThread("bufferHandler");
+        bufferWorker.start();
+        bufferProcessor = new BufferProcessor(
+                lowResExtractor,
+                lowResDecoder,
+                availableInputBuffers,
+                availableOutputBuffers
+        );
+    }
+
+    /**
+     * Shut down all resources allocated to the MotionPhotoReader instance.
+     */
+    public void close() {
+        Log.d("ReaderActivity", "Closed decoder and extractor");
+        try {
+            fileInputStream.close();
+            Log.d("ReaderActivity", "Close file input stream");
+
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR2) {
+            mediaHandler.getLooper().quitSafely();
+            Log.d("ReaderActivity", "Safely quit looper");
+        }
+        else {
+            mediaHandler.getLooper().quit();
+            Log.d("ReaderActivity", "Quit looper");
+        }
+        lowResDecoder.release();
+        lowResExtractor.release();
+        bufferWorker.interrupt();
```

Comment:
this line isn't needed. `bufferHandler` takes ownership of `bufferWorker` via the looper and is properly closed with `bufferHandler.getLooper().quitSafely()`

Reply:
This line actually isn't even used anymore since switching from BufferHandler to BufferProcessor, so I deleted the bufferWorker field entirely (it was a HandlerThread object for handling the previously asynchronous calls that are now synchronous mode).